src/hotspot/share/opto/vectornode.hpp

```diff
@@ -1675,27 +1675,39 @@ class VectorBoxNode : public InlineTypeNode {
   void set_box_type(const TypeInstPtr* box_type) { _box_type = box_type; }
   void set_vec_type(const TypeVect* vec_type) { _vec_type = vec_type; }
 
-  VectorBoxNode(ciInlineKlass* vk, Node* oop, bool null_free, bool is_buffered) :
-    InlineTypeNode(vk, oop, null_free, is_buffered) {}
+  VectorBoxNode(Compile* C, ciInlineKlass* vk, Node* oop, const TypeInstPtr* box_type, const TypeVect* vt, bool null_free, bool is_buffered) :
+    InlineTypeNode(vk, oop, null_free, is_buffered) {
+      init_flags(Flag_is_macro);
+      init_class_id(Class_VectorBox);
+      set_vec_type(vt);
+      set_box_type(box_type);
+      C->add_macro_node(this);
+  }
 
   static VectorBoxNode* make_box_node(PhaseGVN& gvn, Compile* C, Node* box, Node* val,
                                       const TypeInstPtr* box_type, const TypeVect* vt) {
     ciInlineKlass* vk = static_cast<ciInlineKlass*>(box_type->inline_klass());
-    VectorBoxNode* box_node = new VectorBoxNode(vk, box, true, vk->is_empty() && vk->is_initialized());
+    ciInlineKlass* payload = vk->declared_nonstatic_field_at(0)->type()->as_inline_klass();
+
+    Node* payload_oop = payload->is_initialized() ? default_oop(gvn, payload) : gvn.zerocon(T_PRIMITIVE_OBJECT);
+    Node* payload_value = InlineTypeNode::make_uninitialized(gvn, payload, true);
+    payload_value->as_InlineType()->set_field_value(0, val);
+    payload_value = gvn.transform(payload_value);
```

Comment:
Can we move  this to `expand_vbox_node` pass? We can keep the simple two inputs `Box` and `Value` for VectorBox like before. This can simply the change in `merge_through_phi`. 

Besides, `payload_oop` is not used, so can it be removed?

Reply:
Current structure exactly comply with Java class definition, Box classes encapsulate primitive class fields. Existing valhalla routines for buffering and IR creation e.g. _make_from_oop_ etc are all recursive in nature and create an InlineTypeNode for every primitive class object field.